include/rocksdb/utilities/replayer.h

```diff
@@ -0,0 +1,52 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+#ifndef ROCKSDB_LITE
+
+#include <memory>
+
+#include "rocksdb/rocksdb_namespace.h"
+#include "rocksdb/status.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+struct ReplayOptions {
+  uint32_t num_threads;
+  double fast_forward;
+
+  ReplayOptions() : num_threads(1), fast_forward(1.0) {}
+  ReplayOptions(uint32_t num_of_threads, double fast_forward_ratio)
+      : num_threads(num_of_threads), fast_forward(fast_forward_ratio) {}
+};
+
+// Replayer helps to replay the captured RocksDB operations.
+// The Replayer can either be created from DB::NewReplayer method, or be
+// instantiated via db_bench today, on using "replay" benchmark.
+class Replayer {
+ public:
+  virtual ~Replayer() {}
+
+  // Make some preparation before replaying the trace.
+  virtual Status Prepare() { return Status::NotSupported(); }
+
+  // Read one trace and execute it. This function is thread-safe. Trace
+  // timestamps are ignored.
+  // Return Status::OK() if the execution was successful (Get() and MultiGet()
+  // will still return Status::OK() even if they returned Status::NotFound());
+  // Status::Incomplete() if no more trace available;
+  // Status::NotSupported() if the operation is not supported;
+  // Otherwise, return the corresponding error.
+  virtual Status Step() { return Status::NotSupported(); }
+
+  // Replay all the traces from the provided trace stream, taking the delay
+  // between the traces into consideration.
+  virtual Status Replay(ReplayOptions /*options*/ = ReplayOptions()) {
```

Comment:
I prefer to set the options at Prepare()

Reply:
I think we should not provide `Prepare()`, but instead, provide `Reset()`. Given the current API, a user may do a few `Step()`s then start `Replay()`. Or he/she can start `Replay()` and interrupt the process, then start `Step()`, and then start `Replay()`. Considering the where Replayer is used currently (db_bench), we can make the prepare/init step on-demand when a replay/step starts, and hide this step from the user. But the user does need a function to restart if a replay is finished, or the user wants to abort and restart.

In this case, it's strange to set ReplayOptions in `Prepare()`, unless you want 2 sets of options, one for controlling playback speed, such as num threads, fast forward rate, etc.  Another set is explicit used in initialization, for example, control how to read/parse the header or each trace.